### DISCUSSION TITLE
How can we wait for the exit code of a subprocess asynchronously?

### DISCUSSION THREAD

**Discussion by Atry** [facebook/hhvm#8942] [Ideas]
Hi all, I am working on a new HSL API for creating subprocesses. One of the goals is to make the new API completely asynchronous. Unfortunately, currently we don't have a good mechanism to wait for the exit code of a subprocess asynchronously.

I think we could expose a Hack function calling `LightProcess::SetLostChildHandler` to solve the problem if LightProcess is enabled, and provide a fallback implementation to wait for the exit code synchronously when LightProcess is not enabled.

The proposed native function would like this:

``` Hack
function add_lost_child_handler(int $pid, (function():void{}) $handler):void;
function remove_lost_child_handler(int $pid):void;
```

What do you think?

**Comment by fredemmott** [comment]
This will need modifying LightProcess, but it would be much better to natively expose an Awaitable; we generally consider wrapping callbacks to awaitables promise-style to be an antipattern, and it's intentionally awkward to do.

**Reply by fredemmott** [reply]
e.g. something like:

```hack
<<__Native>>
function waitpid_async(int $pid): Awaitable<void>;
```

**Reply by Atry** [reply]
> we generally consider wrapping callbacks to awaitables promise-style to be an antipattern, and it's intentionally awkward to do.

Do we consider wrapping callbacks to awaitables in C++ to be an anitpattern? Because anyway we will have to wrapping callbacks to awaitables somewhere.

**Reply by fredemmott** [reply]
It's necessary in some places in C/C++ (when the library we're using doesn't give us an alternative, and libevent doesn't provide a usable abstraction), but async *Hack* code should always be structured around Awaitables.

**Reply by fredemmott** [reply]
So, there's two options here - and both have good paths that don't require callbacks:

# use LightProcess

LightProcess does the work in subprocesses, and communicates with them over unix sockets.

One way to do this would be to make the new waitpid handler return an FD which it success/error data to and closes when complete. The `__Private` native code could wrap that up in an `OS\FileDescriptor`.

We'd then have something like:

```hack
function waitpid_async(int pid): Awaitable<void> {
  auto $fd = HH\Lib\_Private\_OS\waitpid_fd(pid);
  $result = await OS\poll_async($fd, ...);
  if ($result['exit_code'] == 0) {
    return;
  }
  throw Exception();
}
```

Search hhvm for `afdt_send_fd` (and the wrapper, `cli_write_fd`) for examples of passing FDs between processes.

# Don't do this: use a thread

This is a workable approach for 'I have a non-awaitable API and need an awaitable one' in general, but not here. For reference, look at `curl_multi_await` and the `CurlMultiAwait` classes for examples of doing this.

We shouldn't do this here as the waitpid needs to execute in the process that spawned the process, which isn't necessarily the same process - for example, in CLI server mode, subprocesses are spawned as children of the CLI client, not the HHVM server, so that they are ran with the correct unix permissions.

**Reply by Atry** [reply]
I think we don't need an additional extension to pass an FD to a subprocess. The current `fork_and_exec` support passing arbitrary FDs other than stdin/stdout/stderr. We can simply create a pipe, and pass it to write FD to the child process, then use the read FD to monitor its closure.

**Reply by fredemmott** [reply]
There's two subprocesses here:
- the one we explicitly created with `fork_and_exec`
- the lightprocess worker running `waitpid`, which is also the process that actually ran the `exec()` syscall

hhvm process -> lightprocess worker process -> requested child process

We need a way to receive updates from the second one.

**Reply by Atry** [reply]
How would it affect the approach I described? I suppose if we could pass an arbitrary FD to the "requested child process", we get the notification of the closure of the  "requested child process", then we could simply call the synchronous `waitpid` provided by the lightprocess server.

**Reply by fredemmott** [reply]
- We should be able to spawn the requested child process with *no* attached FDs (though that's a bit weird).
- Some executables also aggressively close FDs on launch to get to a clean state (e.g. lightprocess itself does this) or to reduce the risk of security issues
- It also needs to indicate the status code, not just completion - though, if passing an extra FD to the child /were/ acceptable, this could be resolved by calling waitpid after the handle is closed

**Reply by Atry** [reply]
I think the `HH\Lib\_Private\_OS\waitpid_fd` extension you proposed is called `pidfd_open` https://man7.org/linux/man-pages/man2/pidfd_open.2.html . 

Unfortunately it is only available in Debian 11+. Anyway we can use its name in our polyfill implementation based on SIGCHLD

**Reply by Atry** [reply]
But the proposed `fd` based extension is kind of Linux-biased, because in BSD/macOS, the `EVFILT_PROC` event is directly supported by kqueue and no file descriptor is required to monitor the process exit status.

**Reply by fredemmott** [reply]
We don't need extensions, or to get an FD/awaitable for the process itself; the flow (not the code) of the lightprocess worker currently looks something like this:

```
cmd_and_args = receive_from_hhvm();
pid = fork { exec ( foo) };
send_to_hhvm(pid);
while (next = recv_from_hhvm()) {
  switch(next) {
     // ... other cases ...
     case "waitpid": { // existing blocking thing
        int pid = recv_from_hhvm();
        auto result = waitpid(pid);
        auto saved_errno = errno;
        send_to_hhvm(result);
        if (result < 0) {
          send_to_hhvm(saved_errno);
        }
     }
  }
}
```

I'm suggesting adding something like:

```
case "waitpid_async":
  auto pid = recv_from_hhvm();
  auto [r, w] = pipe();
  // initial function call in C++ returns the FD or awaitable here
  send_fd_to_hhvm(r);
  // hhvm is not waiting for anything below until an `await` happens
  close(r);
  auto result = waitpid(pid);
  auto saved_errno = errno;
  // awaitable maybe resolves here...
  dprintf(w, "%d\n", result);
  // ... and here... 
  if (result < 0) {
    dprintf(w, "%d\n", saved_errno);
  }
  //... and here; or, only here if we wrap it natively rather than returning an OS\FileDescriptor to hack code
  close(w);
```

**Reply by Atry** [reply]
Thank you. I supposed that `waitpid_async` could be named `pidfd_open` if it returns an fd, and we can implement it with the help of a `SIGCHLD` handler instead of blocking a lightprocess worker process.

**Reply by fredemmott** [reply]
I'd rather not name it the same as a syscall (even if linux-specific), especially when it doesn't do quite the same thing: pidfd_open attaches to an existing FD, this can only work from the parent of the process that spawned it - which does mean we still need to do it in the lightprocess worker that originally spawned it, like we do for blocking waitpid

**Reply by Atry** [reply]
I didn't find that `pidfd_open` is limited to child processes. Any way we cannot depend on it right now because it is Linux specific and too new to be available in our supported distributions.

**Reply by fredemmott** [reply]
pidfd_open is not limited to child processes

The fd approach I suggested *is*, so shouldn't be called pidfd_open, and needs to be in the original worker thread

**Comment by Atry** [comment]
Shall we expose some functions of `LightProcess` to Hack, like:

``` hack
newtype LightProcess = mixed;

// General LightProcess functions
<<__Native>>
function select_read_async(LightProcess $proc): Awaitable<void>;
<<__Native>>
function release_light_process(LightProcess $proc): void;
<<__Native>>
function acquire_light_process(): LightProcess;


// waitpid functions
newtype stat_loc = int;
<<__Native>>
function write_waitpid(LightProcess $proc, pid_t $pid, int $options = 0): void;
<<__Native>>
function read_waitpid_result(LightProcess $proc): stat_loc; 
```

The current implementation of `LightProcess::waitpid` can be split into `write_waitpid` and `read_waitpid_result`, where `write_waitpid` includes the `lwp::write` part and `read_waitpid_result` includes the `lwp::read` part. With the help of general `LightProcess` functions to select the `afdt_fd`, the `waitpid_async` can be implemented roughly as following:

``` hack
async function waitpid_async(pid_t $pid, int $options = 0): stat_loc {
  $proc = acquire_light_process();
  try {
    write_waitpid($proc, $pid, $options);
    await select_read_async($proc);
    return read_waitpid_result($proc);
  } finally {
    release_light_process($proc);
  }
}
```
